### DISCUSSION TITLE
Callback functions for Trainers ?

### DISCUSSION THREAD

**Discussion by AxKo** [Toni-SM/skrl#167] [Q&A] [answered]
Hi Everybody,

I recently discovered skrl and I like the coding style very much.
I wonder if it is possible to use callbacks with SequentialTrainer (I didn't find any documentation) ?
Specifically I like to write my own callback for progress monitoring (I don't like Tensorboard so much).

Many thanks

**Comment by Toni-SM** [comment, marked as answer]
Hi @AxKo 

Currently there are no callbacks in the library implementation. 
However, a possible solution for a custom progress monitoring is to overwrite the [Agent.write_tracking_data](https://skrl.readthedocs.io/en/latest/api/agents.html#skrl.agents.torch.base.Agent.write_tracking_data) (see [implementation](https://github.com/Toni-SM/skrl/blob/636936f3ac49c6d2260bd130d72b789ca6dfe42b/skrl/agents/torch/base.py#L192)) method to handle the `self.tracking_data` content according to the desired implementation.

```python
from skrl.agents.torch.ppo import PPO

class DerivedPPO(PPO):
    def write_tracking_data(self, timestep: int, timesteps: int) -> None:
        # process data for custom progress monitoring
        #===================================
        for k, v in self.tracking_data.items():
            pass
        #===================================
            
        # reset data containers for next iteration
        self._track_rewards.clear()
        self._track_timesteps.clear()
        self.tracking_data.clear()

# ...

agent = DerivedPPO(...)
```
